Checkpoint enchancements

pbehne opened the pull request:
This PR provides enhancements to the existing automatic wall time checkpoint system. Specifically,

- The `Outputs/wall_time_checkpoint` shortcut syntax is introduced as an easy way to turn off automatic checkpoint output.
- An informative block is output to the MOOSE header describing the configuration of the checkpoint system for each simulation.
- A new integration test is added to verify that subapps are not auto-checkpointed if not requested.
- A new regression test is added that sets the checkpoint wall time interval to a value smaller than the time required to write a checkpoint to verify that no issues are encountered.
- The default checkpoint wall time interval is adjusted from 10 minutes to 1 hour based on community feedback.
- The checkpoint format is changed from binary (.cpr) to compressed ascii (.cpa.gz) so it will run on systems without XDR support.

Closes #27240, #27205, #26682

Related PRs to update apps:
Griffin: https://github.inl.gov/ncrc/griffin/pull/1876
Bison: https://github.inl.gov/ncrc/bison/pull/5994

pbehne committed:
Added option to disable wall time checkpoints

This option may be set using the checkpoint block input syntax or using a
shortcut syntax ‘Outputs/disable_wall_time_checkpoints=true’. In addition, the
‘Outputs/checkpoint=true’ shortcut syntax was moved from CommonOutputAction
into AutoCheckpointAction. This has the benefit that any automatic and shortcut
checkpoints are simply created in the same place instead of in multiple places.
Ref #27240

pbehne committed:
Added test to make sure wall time checkpoints aren't created for subapps if not requested.
Ref #27240

pbehne committed:
Changed default checkpoint wall time interval from 10 minutes to 1 hour.
Ref #27240

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2157571/) on 847f743 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27268/site)

This comment will be updated on new commits.

pbehne committed:
Patch to ensure test-checkpoint-half-transient stuff is only done at the master app level.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2159072/) on d1c40a2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b02bb1</th>
      <th colspan=4>#27268 d1c40a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.15%</td>
      <td>85.15%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>102638</td>
      <td>102651</td>
      <td>+13</td>
      <td>30</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17899</td>
      <td>17898</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27268/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27268/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27268/coverage/xfem)

</details>



This comment will be updated on new commits.

pbehne committed:
Addressed SQA issues.

pbehne committed:
Changed checkpoint format from binary to ascii.

pbehne committed:
Updated failing test.

pbehne committed:
Updated src code for failing prefix test (attempt 2).

pbehne committed:
Added Ref #27240 to updated tests.

loganharbour commented:
Tag when needed for a review, silencing until then

pbehne committed:
Changed remaining instances of cpr to cpa, fixed another test.

pbehne committed:
Changed default file suffix from cpr to cpa when splitting, updated relevant tests.

pbehne committed:
Regenerating split mesh file on linux and restricting the platform per Roy.

pbehne committed:
Updated another test.

pbehne committed:
Fixed yet another test...

pbehne commented:
@loganharbour, ready for review. I'll work on fixes for the failing tests in Bison and Griffin in the meantime. Should be trivial.

roystgnr commented:
Not going all the way to .cpa.gz yet?

loganharbour commented:
> Not going all the way to .cpa.gz yet?

please clarify?

roystgnr commented:
The missing context here was some discussion on Slack - https://moosedevelopers.slack.com/archives/C01054VRUEM/p1712693072807749?thread_ts=1711666351.143939&cid=C01054VRUEM

The summary is that IMHO there's no reason to prefer binary .cpr over ASCII+gzip'ed .cpa.gz  - the latter is usually a little smaller, it's more compatible, and it's easier to debug with.  But .cpr vs .cpa has tradeoffs in both directions; the .cpa files will generally be significantly larger.  So right now I'm wondering if this PR (which on first skim is just switching to .cpa?) is one step in the .cpa.gz direction, or if there's some reason to avoid gzip that I hadn't thought about, or (most seriously) if Patrick tried gzip but something didn't work correctly.

dschwen commented:
Are all walltime checkpoints kept? When I do manual checkpointing I usually only keep the last 2 checkpoints. For long running simulations the checkpoints can occupy massive amounts of disk space, especially if we move to uncompressed ASCII checkpoints. I would not want my simulations fail because they run out of disk space.

pbehne commented:
By default, only the last 2 wall time checkpoints are kept, just like the default for other checkpoints. This can be changed through the same `num_files` parameter.

pbehne commented:
> Not going all the way to .cpa.gz yet?

I misunderstood our discussion on slack in that I thought that compression and decompression is performed automatically when using cpa format. What library should I use to do this, zlib?

roystgnr commented:
It's "automatic" but based on filename.  Ask for a `.cpa` extension and you get plain ASCII; ask for `.cpa.gz` and you get gzipped ASCII.

pbehne committed:
Changed instances of .cra to .cpa.gz to enable (de)compression, updated tests.

pbehne committed:
Updated test input mesh from cpa to cpa.gz.

pbehne committed:
Fixed failing tests.

pbehne committed:
Regenerated splits on rod so they will run on civet. Marked tests as linux only.

pbehne committed:
Renamed 'disable_wall_time_checkpoints' param to 'wall_time_checkpoint'.

pbehne committed:
Fixed yet another failing test...

pbehne committed:
Fixed another test, updated param name in documentation file.

loganharbour commented:
> It's "automatic" but based on filename. Ask for a `.cpa` extension and you get plain ASCII; ask for `.cpa.gz` and you get gzipped ASCII.

@pbehne was this change made? This is pretty significant and I'm not sure it'll work 100% with all of the automated file searching we do when specifying cp directories

pbehne commented:
> > It's "automatic" but based on filename. Ask for a `.cpa` extension and you get plain ASCII; ask for `.cpa.gz` and you get gzipped ASCII.
> 
> @pbehne was this change made? This is pretty significant and I'm not sure it'll work 100% with all of the automated file searching we do when specifying cp directories

Yes, the change has been made.

loganharbour commented on framework/src/actions/AutoCheckpointAction.C:
```diff
@@ -25,46 +33,42 @@ AutoCheckpointAction::AutoCheckpointAction(const InputParameters & params) : Act
-    // Get most recently added Checkpoint object and error
+  const bool shortcut_syntax = getParam<bool>("checkpoint");
+
+  if (num_checkpoints > 1 || (num_checkpoints == 1 and shortcut_syntax))
```
```
  if (num_checkpoints > 1)
    checkpoints[0]->mooseError("Multiple Checkpoint objects are not allowed and there is more than one Checkpoint defined in the 'Outputs' block.");
  if (num_checkpoints == 1 && shortcut_syntax)
    paramError("checkpoint", "Shortcut checkpoint syntax cannot be used with another Checkpoint object in the 'Outputs' block");
```

Adds more context to the errors (input positions, etc)

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -104,6 +108,16 @@ Checkpoint::Checkpoint(const InputParameters & parameters)
   // 'The --output-wall-time-interval parameter is necessary for testing
   // and should only be used in the test suite.
   Output::setWallTimeIntervalFromCommandLineParam();
+  // We want to do this here and not in the constructor so it overrides --output-wall-time-interval
+  if (!getParam<bool>("wall_time_checkpoint"))
+  {
+    _wall_time_interval = std::numeric_limits<Real>::max();
+    mooseInfo("Wall time checkpoints are disabled.");
+  }
+  else
+    mooseInfo(
+        "Checkpoints will be written every " + std::to_string(_wall_time_interval) +
+        " seconds of wall time or at the end of the current timestep, whichever occurs later.");
```
Adding a new moose info print to _nearly every run_ isn't a practice we like to do too much. These should be removed

pbehne replied:
Could you please post this comment on #27205?

loganharbour replied:
Commented

loganharbour commented on test/tests/mesh/checkpoint/tests:
```diff
@@ -10,18 +10,22 @@
     cli_args = 'Outputs/file_base=test_2 --use-split --split-file checkpoint_split_in'
     max_parallel = 2
     min_parallel = 2
+    # checkpoint files are platform-dependent
```
Removing _every_ test for mac here is dangerous. do we not still support ascii checkpointing?

pbehne replied:
Before this PR we only supported binary (cpr) checkpoints. Now we only support compressed ascii (cpa.gz) checkpoints.

The reason for adding the platform specification here is that this test has to read mesh files that were generated on a linux platform and stored as part of the moose repo - they are not generated in a prerequisite test.

See below conversation with @roystgnr that prompted me to add patform dependence:

>Me: When we default to using ASCII instead of binary to store checkpoints, does this give us the benefit that a checkpoint written on one architecture/OS can be read and used on a different architecture/OS? Or do we expect there to be differences in the compression/decompression on different architectures/OSs that cause issues?

> Roy: Both ASCII and binary can be architecture independent.  The whole point of XDR is to handle things like endianness differences.  But in the specific case of .cpa files... on the one hand, you'd get a little extra compatibility, because IIRC we have a bunch of integer-size stuff in .xdr headers that isn't in .cpr, and on the other hand no, because half the point of CheckpointIO was to have a non-format where we don't need 10 layers of compatibility, and if you try to reload one after so much as a git pull since saving it our only guarantee is that we'll laugh at you.

That being say, an easy way to have this tests on all platforms would to be to add a test that generates the split mesh instead of storing the mesh files as part of the repo.

roystgnr replied:
I just dug into checkpoint_io.C, and it looks like I actually *did* get around to putting the integer-size stuff into .cpr ... or rather, Derek forked CheckpointIO off of XDRIO after I added it to .xdr?

Anyway, we've probably actually maintained architecture independence since then; now that I think about it, we added some checked-in .cpr files to the MOOSE test suite back when I was working on a feature for it ... possibly N-to-M restarts?

pbehne replied:
Yeah, there seemed to be architecture independence for this particular test while the files were cpr, but as cpa.gz generated on linux, I get the following error message on mac:

`mesh/checkpoint.test_2: *** ERROR ***
mesh/checkpoint.test_2: While reading restartable data in "/Users/behnpa/projects/moose_checkpoint/test/tests/mesh/checkpoint/checkpoint_split_in.cpa.gz/meta_data_mesh.rd":
mesh/checkpoint.test_2: 
mesh/checkpoint.test_2: The backup is not compatible
mesh/checkpoint.test_2: 
mesh/checkpoint.test_2: The hash code check for a basic type (int) failed.
mesh/checkpoint.test_2: It is likely that this backup was stored with a different architecture or operating system`

roystgnr replied:
...

If anything .cpa.gz should be more independent.

That error message isn't coming from CheckpointIO, though.  The .cpa.gz files are being written/read to actual files, right?  Not interleaved into some stream where a bad read could move on to non-CheckpointIO data too early or too late?

pbehne replied:
These are the files I'm seeing:

`(moose) INL654724:checkpoint behnpa$ ls -R checkpoint_split_in.cpa.gz
2			4			8			meta_data_mesh.rd

checkpoint_split_in.cpa.gz/2:
header.gz	split-2-0.gz	split-2-1.gz

checkpoint_split_in.cpa.gz/4:
header.gz	split-4-0.gz	split-4-1.gz	split-4-2.gz	split-4-3.gz

checkpoint_split_in.cpa.gz/8:
header.gz	split-8-1.gz	split-8-3.gz	split-8-5.gz	split-8-7.gz
split-8-0.gz	split-8-2.gz	split-8-4.gz	split-8-6.gz

checkpoint_split_in.cpa.gz/meta_data_mesh.rd:
data	header`

In my experience, when `ls` is run on a compressed file, the command can't see into the archive. Why is that not the case here?

pbehne replied:
I just recompiled without the hash code checks, and all the exodiff tests passed. I'll look into the check a bit more and get with Roy if I can't figure it out.

pbehne replied:
The check seems simple enough at the surface, I'm handing the investigation off to @loganharbour as to whether this check is needed.

loganharbour replied:
The hash code checks exist as a sanity check to the user to make sure that they're not running a binary checkpoint from another machine. They run typeid(T) on the type being stored and make sure it's the same on the other end.

pbehne replied:
Here is the resolution that I will implement that we discussed in the NEAMS meeting:

Instead of storing the split mesh files on github, these will be generated on-the-fly in prerequisite tests.

pbehne commented on framework/src/meshgenerators/FileMeshGenerator.C:
```diff
@@ -140,11 +140,11 @@ FileMeshGenerator::deduceCheckpointPath(const MooseObject & object, const std::s
-  // xxxx_mesh.cpr -> xxxx-mesh.cpr
-  const std::string old_ending = "_mesh.cpr";
+  // xxxx_mesh.cpa.gz -> xxxx-mesh.cpa.gz
+  const std::string old_ending = "_mesh.cpa.gz";
   if (std::equal(old_ending.rbegin(), old_ending.rend(), file_name.rbegin()))
   {
-    const std::string new_ending = "-mesh.cpr";
+    const std::string new_ending = "-mesh.cpa.gz";
```
If we are completely dropping .cpr support, I think we can also drop this prefix renaming support. @loganharbour?

pbehne committed:
Addressed most of Logan's review.

pbehne committed:
Cleaned up Checkpoint constructor, added execute_on values to header print-out.

pbehne committed:
Changed duplicate requirement.

loganharbour commented:
Can you give examples of the header output? It's a bit more difficult to follow

pbehne commented:
> Can you give examples of the header output? It's a bit more difficult to follow

See attached for more examples.

`outputs/checkpoint.interval/test_files: Current Time:            Wed Apr 24 20:44:09 2024
outputs/checkpoint.interval/test_files: Executable Timestamp:    Tue Apr 23 10:30:46 2024
outputs/checkpoint.interval/test_files: 
outputs/checkpoint.interval/test_files: Checkpoint:
outputs/checkpoint.interval/test_files:   Wall Time Interval:      Every 3600.000000 s
outputs/checkpoint.interval/test_files:   User Checkpoint:         Outputs/out
outputs/checkpoint.interval/test_files:   # Checkpoints Kept:      2
outputs/checkpoint.interval/test_files:   Execute On:              TIMESTEP_END 
outputs/checkpoint.interval/test_files: 
outputs/checkpoint.interval/test_files: Parallelism:
outputs/checkpoint.interval/test_files:   Num Processors:          1
outputs/checkpoint.interval/test_files:   Num Threads:             1
outputs/checkpoint.interval/test_files: 
outputs/checkpoint.interval/test_files: Mesh: 
outputs/checkpoint.interval/test_files:   Parallel Type:           replicated
outputs/checkpoint.interval/test_files:   Mesh Dimension:          2
outputs/checkpoint.interval/test_files:   Spatial Dimension:       2
outputs/checkpoint.interval/test_files:   Nodes:                   121
outputs/checkpoint.interval/test_files:   Elems:                   100
outputs/checkpoint.interval/test_files:   Num Subdomains:          1
`

`outputs/checkpoint.default/recover: Executable Timestamp:    Tue Apr 23 10:30:46 2024
outputs/checkpoint.default/recover: 
outputs/checkpoint.default/recover: Checkpoint:
outputs/checkpoint.default/recover:   Wall Time Interval:      Every 3600.000000 s
outputs/checkpoint.default/recover:   User Checkpoint:         Disabled
outputs/checkpoint.default/recover:   # Checkpoints Kept:      2
outputs/checkpoint.default/recover:   Execute On:              TIMESTEP_END 
outputs/checkpoint.default/recover: 
outputs/checkpoint.default/recover: Parallelism:
outputs/checkpoint.default/recover:   Num Processors:          1
outputs/checkpoint.default/recover:   Num Threads:             1
`

`outputs/checkpoint.default/wall_time_interval: 
outputs/checkpoint.default/wall_time_interval: Checkpoint:
outputs/checkpoint.default/wall_time_interval:   Wall Time Interval:      Every 0.020000 s
outputs/checkpoint.default/wall_time_interval:   User Checkpoint:         Disabled
outputs/checkpoint.default/wall_time_interval:   # Checkpoints Kept:      2
outputs/checkpoint.default/wall_time_interval:   Execute On:              TIMESTEP_END 
outputs/checkpoint.default/wall_time_interval: 
outputs/checkpoint.default/wall_time_interval: Parallelism:
`

`outputs/checkpoint.default/wall_time_interval_disabled: Current Time:            Wed Apr 24 20:44:10 2024
outputs/checkpoint.default/wall_time_interval_disabled: Executable Timestamp:    Tue Apr 23 10:30:46 2024
outputs/checkpoint.default/wall_time_interval_disabled: 
outputs/checkpoint.default/wall_time_interval_disabled: Checkpoint:
outputs/checkpoint.default/wall_time_interval_disabled:   Wall Time Interval:      Disabled
outputs/checkpoint.default/wall_time_interval_disabled:   User Checkpoint:         Disabled
outputs/checkpoint.default/wall_time_interval_disabled: 
outputs/checkpoint.default/wall_time_interval_disabled: Parallelism:
outputs/checkpoint.default/wall_time_interval_disabled:   Num Processors:          1
outputs/checkpoint.default/wall_time_interval_disabled:   Num Threads:             1
`

[output.txt](https://github.com/idaholab/moose/files/15103317/output.txt)

pbehne commented:
The copy and past into code blocks above really messes with the indent. I recommend downloading and catting `output.txt`.

pbehne committed:
Removed dependence on platform, split files now generated on the fly instead
of stored, added *.cpa.gz to .gitignore.

pbehne committed:
Set recover to false for tests that generate mesh splits according to the precedent set in mesh/splitting.

pbehne committed:
Don't run splitting tests in replicated mode.

pbehne commented:
@loganharbour, I think we are good now. Bison is passing at https://github.inl.gov/ncrc/bison/pull/5933 and I'm currently working on Griffin.

loganharbour commented on framework/src/actions/AutoCheckpointAction.C:
```diff
@@ -25,46 +33,45 @@ AutoCheckpointAction::AutoCheckpointAction(const InputParameters & params) : Act
-      // autosave
-      checkpoints[0]->setAutosaveFlag(CheckpointType::USER_CREATED);
-    }
+    cp_params.set<bool>("_built_by_moose") = true;
```
this is just an aside as this is old code - but you can actually query the action system for the current task name. that is, in the checkpoint object you can see if the current task is `auto_checkpoint_action` and by that determine whether or not this is the automatic one.

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -248,3 +230,69 @@ Checkpoint::updateCheckpointFiles(CheckpointFileNames file_struct)
+    valid_execute_on_value = {EXEC_INITIAL, EXEC_TIMESTEP_END, EXEC_FINAL};
+    valid_execute_on_values[6] = valid_execute_on_value;
+  }
+
```
```
const auto & execute_on = getParam<ExecFlagEnum>("execute_on");
const std::set<ExecFlagType> allowed = {EXEC_INITIAL, EXEC_TIMESTEP_END, EXEC_FINAL};
for (const auto & value : execute_on)
  if (!allowed.count(value))
    paramError("execute_on", "The exec flag ", value, " is not allowed... etc");
```

I know that this is old code too, but this is probably the better option

pbehne replied:
Beautiful.

loganharbour commented on framework/include/outputs/Checkpoint.h:
```diff
@@ -83,6 +82,12 @@ class Checkpoint : public FileOutput
   void setAutosaveFlag(CheckpointType flag) { _checkpoint_type = flag; }
 
+  /// Gets information about the checkpoint
+  const std::vector<std::pair<std::string, std::string>> & getCheckpointInfo() const
```
Few things:

- Reduce the need to record it. Just call it on the fly in the header, this reduces the need to do the set/get thing. This isn't expensive.
- Don't make it vec<pair<...>>, just do a vector of the preformatted things that you can loop to and explain in the docstring what it'll look like

pbehne replied:
Please see my changes and resolve this conversation if you are satisfied.

loganharbour commented on test/tests/mesh/checkpoint/tests:
```diff
@@ -1,33 +1,57 @@
 [Tests]
```
Let's just remove the duplicate tests. These requirements aren't particularly useful, anyway. Testing this with just a single set of procs is fine. Requirements should be general enough s.t. they don't list number of procs

loganharbour reviewed (changes requested):
Last set. Everything is viewed. Good to go in after this!

pbehne committed:
Addressed Logan's review.

pbehne committed:
Addressed Logan's review pt 2.

pbehne committed:
Updated expected std::str.find position to take into account different length of new .cpa.gz extension compared to old .cpr extension.

pbehne committed:
Changed cpr extension to cpa.gz, added meta_data_mesh.rd files to check_files.

pbehne committed:
Updated documentation.

pbehne committed:
Removed deprecated prefix detection as it is no longer useful.

pbehne committed:
Forgot to save file removing deprecated preix tests.

loganharbour commented on framework/include/outputs/Checkpoint.h:
```diff
@@ -107,4 +119,7 @@ class Checkpoint : public FileOutput
   std::deque<CheckpointFileNames> _file_names;
+
+  /// Vector of pairs describing checkpoint structure
+  std::vector<std::pair<std::string, std::string>> _checkpoint_info;
```
Isn't this unused now?

pbehne replied:
Correct.

pbehne committed:
Removed unused attribute.

roystgnr commented on framework/src/actions/SetupMeshAction.C:
```diff
@@ -205,17 +205,17 @@ std::string
   std::string split_file_ext;
   auto pos = split_file.rfind(".");
   if (pos != std::string::npos)
     split_file_ext = split_file.substr(pos + 1, std::string::npos);
```
Wait, how is this still going to find the full extension if the extension is ".cpa.gz"?  Won't the `rfind(".")` stop at the second "." and just leave `split_file_ext == "gz"`, which would then fail to hit the code path we expect below?

pbehne replied:
Good catch. I'm surprised nothing in the test suite caught this. I'll add a test.

roystgnr replied:
Right!?  If I'd been more confident and less confused I'd have requested such a test myself, but I still had some hope that you'd come back with, "No, what you're missing is that this code still works because ..."

roystgnr replied:
I like your solution here; it has the issue of false positives in the *other* direction (directories with `.` in the name, say), but I don't know how we'd avoid that without a predefined list of possible suffixes or a short maximum suffix length or something.

pbehne replied:
We could always do an rfind("/") to detect any folders prepended to the file name and split off the file part. But that's an issue for future moose developers, not current moose developers :)

roystgnr commented on framework/src/actions/SplitMeshAction.C:
```diff
@@ -59,28 +59,28 @@ SplitMeshAction::act()
     if (split_file_arg_ext == "cpr")
       checkpoint_binary_flag = true;
-    else if (split_file_arg_ext == "cpa")
+    else if (split_file_arg_ext == "cpa.gz")
```
Same issue here - if we had "foo.cpa.gz", won't `split_file_arg_ext == "gz"`?

roystgnr commented on framework/src/actions/SplitMeshAction.C:
```diff
@@ -59,28 +59,28 @@ SplitMeshAction::act()
   if (fname == "")
     fname = split_file_arg;
-  fname = MooseUtils::stripExtension(fname) + (checkpoint_binary_flag ? ".cpr" : ".cpa");
+  fname = MooseUtils::stripExtension(fname) + (checkpoint_binary_flag ? ".cpr" : ".cpa.gz");
```
Same issue here, if we're taking `.cpa.gz` as input too for some reason; `MooseUtils::stripExtension()` also just does `rfind(".")`

pbehne replied:
@roystgnr, are we good to resolve this one? Note that I changed the implementation of `stripExtension` to use find by default.

roystgnr replied:
Sorry I missed that; thanks for double-checking.

pbehne committed:
Created extension getter util that uses find

Consolodation two code snippets that use rfind(".") to get file name extensions
into a utility. Changed the logic to use find(".") by default instead of rfind
to handle to compressed ascii extension tha itself has a dot in it (".cpa.gz").
Changed stripExtension to use find instead of rfind.

pbehne commented:
Based on my changes to address @roystgnr's review, I pushed the latest changes to my BISON and griffin PRs. These are now failing. I'll look more into them tomorrow to see what implications `find` vs `rfind` have.

loganharbour commented:
All good @roystgnr ?

pbehne commented:
Okay, everything is ready on my end.

loganharbour commented:
https://civet.inl.gov/job/2281545/ this is a valid failure I believe

pbehne commented:
> https://civet.inl.gov/job/2281545/ this is a valid failure I believe

"It runs on my machine" :)

I'll invalidate it, and if it keeps failing, investigate more.

loganharbour commented:
> > https://civet.inl.gov/job/2281545/ this is a valid failure I believe
> 
> 
> 
> "It runs on my machine" :)
> 
> 
> 
> I'll invalidate it, and if it keeps failing, investigate more.

Odds are you're missing a dependency in the tests. A lot of skips exist on the installed tests, which leads to a significantly different dependency tree. Regardless, invalidating over and over isn't the solution. Unfortunately you need to actually dig into it.

pbehne committed:
Now strip directory from file name in getExtension to avoid issues with directory names containing dots.

loganharbour reviewed (approved):
@roystgnr I'd appreciate your sanity check one last time

pbehne commented:
@loganharbour, I think we're finally good. I updated my bison and griffin PRs to reflect the latest changes, and those are both passing.

roystgnr reviewed (approved):
I still worry about about foo.bar.e.gz sorts of filenames, but I don't think there's any way to disambiguate "this is the extension" vs "this is just a period in the middle of the filename" without using a predefined set of allowed extensions.

loganharbour commented:
> I still worry about about foo.bar.e.gz sorts of filenames, but I don't think there's any way to disambiguate "this is the extension" vs "this is just a period in the middle of the filename" without using a predefined set of allowed extensions.

We can actually do something akin to the `file` command in linux to get more information about the file, but that's for another day.

pbehne commented:
@lindsayad NEAMS MP project
